Riot/Managers/PushNotification/PushNotificationService.m

```diff
@@ -918,44 +918,75 @@ - (void)handleNotificationInlineReplyForRoomId:(NSString*)roomId
 
     NSArray* mxAccounts = [MXKAccountManager sharedManager].activeAccounts;
 
-    MXKRoomDataSourceManager* manager;
+    __block MXKRoomDataSourceManager* manager;
+    dispatch_group_t group = dispatch_group_create();
 
     for (MXKAccount* account in mxAccounts)
     {
-        MXRoom* room = [account.mxSession roomWithRoomId:roomId];
-        if (room)
+        void(^storeDataReadyBlock)(void) = ^{
+            MXRoom* room = [account.mxSession roomWithRoomId:roomId];
+            if (room)
+            {
+                manager = [MXKRoomDataSourceManager sharedManagerForMatrixSession:account.mxSession];
+            }
+        };
+        
+        if (account.mxSession.state >= MXSessionStateStoreDataReady)
         {
-            manager = [MXKRoomDataSourceManager sharedManagerForMatrixSession:account.mxSession];
+            storeDataReadyBlock();
             if (manager)
             {
                 break;
             }
         }
+        else
+        {
+            dispatch_group_enter(group);
+            
+            //  wait for session state to be store data ready
+            id sessionStateObserver = nil;
+            sessionStateObserver = [[NSNotificationCenter defaultCenter] addObserverForName:kMXSessionStateDidChangeNotification object:account.mxSession queue:nil usingBlock:^(NSNotification * _Nonnull note) {
+                if (manager)
+                {
+                    [[NSNotificationCenter defaultCenter] removeObserver:sessionStateObserver];
+                    return;
```

Comment:
Should we call `dispatch_group_leave(group);` here ?

Reply:
I thought it too, but it caused a crash, because we call `dispatch_group_leave(group)` twice then. It seems `-[NSNotificationCenter removeObserver:]` is async in its own. Because the block is called sometimes after removing observer.

But i'm sure we're not missing a `dispatch_group_leave` here. It seems safe to me.

Reply:
ok. I see.